### PR TITLE
3D implementation #509

### PR DESCRIPTION
**Pull request by BoyaoZhang**
I hope this time it works well. :)

### CODE REVIEW

**Comment by berndbischl** [comment]
hi @zmjones @MasonGallo 

boyao here implemented created some code to have plotLearnerPrediction work in 3d. 
can you help me in review, so we can better tell him to bring this in a form so we can merge this at some point.

@BoyaoZhang:
you also cannot include your subfolder  "3dplots" here. from these examples, you have to create unit tests and possibly a smaller text in the tutorial (for the tutorial: dont do this now and we can help with that)

also you have to watch the travis CI button. if this fails, we cannot merge this.

**Comment by berndbischl** [comment]
@BoyaoZhang 
you need to run R CMD check on your side on the code. the output must be completely clean then. without any notes or complaints. do you know how this works?

**Comment by BoyaoZhang** [comment]
@berndbischl 
I think I've already know how it works, use `devtools::check()`, doesn't it?
It is my first time to pull a request and write package functions. So it will take me some time to learn these basic knowledge. Thank you for your patient. :)

**Comment by berndbischl** [comment]
@BoyaoZhang 
boyao, why are you removing all of the comments from your src code? that makes really no sense and make the code worse not better?
and yes, you could run devtools::check

**Comment by BoyaoZhang** [comment]
@berndbischl 
ehh..... you said that "without any notes or complaints..". So I thought that you want me to remove the comments. I will add them later.

**Comment by MasonGallo** [comment]
@BoyaoZhang this stuff is always hard the first few times - don't get discouraged!

**Comment by berndbischl** [comment]
> ehh..... you said that "without any notes or complaints..". So I thought that you want me to remove the comments. I will add them later

i of course meant the NOTES, ERROS and WARNING that R CMD CHECK produces. 
(which are also very visible on travis now)
not your comments in the code!

**Comment by BoyaoZhang** [comment]
@berndbischl 
Bernd, in the R CMD CHECK I got the following NOTE. I don't know how to fix them. Can you give me some tipps?

```
plotPartialPredictionPlotly: no visible binding for global variable ‘x’
plotPartialPredictionPlotly: no visible binding for global variable ‘y’
plotPartialPredictionPlotly: no visible binding for global variable ‘z’
plotPartialPredictionPlotly: no visible global function definition for
  ‘.’
Undefined global functions or variables:
  . x y z
```

File: R/plotPartialPredictionPlotly.R [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+                  type = "surface")
+    
+    plt = plt %>% layout(scene = list(xaxis = list(title = paste("x: ", x1n, sep = "")),
+                                      yaxis = list(title = paste("y: ", x2n, sep = "")), 
+                                      zaxis = list(title = paste("z: ", target, sep = ""))))
+  }
+  
+  if (obj$task.desc$type == "classif") {
+    grid.dcast.tmp = obj$data
+    grid.dcast.tmp = plyr::ddply(grid.dcast.tmp, .(get(x1n), get(x2n)), function (x) {x$.id = 1:nrow(x); x})
```
Comment (MasonGallo) [inline R/plotPartialPredictionPlotly.R]: I don't have time to run your code, but this line doesn't look right

**Comment by BoyaoZhang** [comment]
@MasonGallo 
Thank you for the advise. I found the problem. I haven't finish writing this Partial Prediction function. And I will add it some days later.

**Comment by BoyaoZhang** [comment]
@berndbischl 
In fact I want to pass data into a list. x, y in the list contain the coordinates ranges, and z contains a grid matrix that used to plot 3D graphics. 
At first I define the list through
`list(x = xxx, y = yyy, z = matrix)`
I solved it through
`list("x" = xxx, "y" = yyy, "z" = matrix)`
I hope it is a good solution.

**Comment by BoyaoZhang** [comment]
@berndbischl @MasonGallo 
hi, can you help me to find what is wrong in my code? There are no errors or notes in my local check.

<code>
** R

Error in .install_package_code_files(".", instdir) : 

files in '/tmp/RtmpFdA41F/Rbuild24f0605fb815/mlr/R' missing from 'Collate' field:

  plotLearnerPredictionPlotly.R

  plotPartialDependencePlotly.R

ERROR: unable to collate and parse R files for package ‘mlr’

* removing ‘/tmp/RtmpFdA41F/Rinst24f0373557de/mlr’

      -----------------------------------

ERROR: package installation failed
</code>

**Comment by berndbischl** [comment]
sorry, this was due to a recent change in mlr master and our fault.

it is fixed now. pls rebase on master and the error will go away

**Comment by berndbischl** [comment]
@zmjones 
dear plotmaster can you please also review this?

**Comment by BoyaoZhang** [comment]
@berndbischl 
I' sorry, that I still need some time to change the codes, because there are some important changes in the new dev version of plotly package. For example the parameter "group" in plot_ly() function, that I used in my functions, has been deprecated. It is more about the bounding region for classification problem.
When I try to use the current version of plotly, there are still some bugs. But the dev version is still not published. Which one should I use?

**Comment by BoyaoZhang** [comment]
@berndbischl 
I prefer the dev version. Maybe you should merge the functions until the dev version is published. But this helps to give a good version of my bachelor thesis. How do you think?

**Comment by berndbischl** [comment]
@BoyaoZhang 
if you can work better with the dev version, stick to that.

**Pull request closed without merging**